checkbox-ng/checkbox_ng/launcher/master.py

```diff
@@ -166,6 +166,41 @@ def invoked(self, ctx):
         else:
             print(_("\nConnection timed out."))
 
+    def check_remote_api_match(self):
+        """
+        Check that agent and controller are running on the same
+        REMOTE_API_VERSION else exit checkbox with an error
+        """
+        agent_api_version = self.sa.get_remote_api_version()
+        controller_api_version = RemoteSessionAssistant.REMOTE_API_VERSION
+        if agent_api_version == controller_api_version:
+            return
+        template_msg = (
+            "The controller that you are using is {} than the agent "
+            "you are trying to connect to.\n"
+            "To solve this, upgrade the {} to the "
+            "{} version.\n"
+            "If you are unsure about the nomenclature see:\n"
+            "https://checkbox.readthedocs.io/en/latest/reference"
+            "/glossary.html\n\n"
+            "Error: (Agent version: {}, Controller version {})"
+        )
+        if controller_api_version > agent_api_version:
+            problem = "newer"
+            solution = ("agent", "controller")
+        else:
+            problem = "older"
+            solution = ("controller", "agent")
+        problem_msg = template_msg.format(
+            problem,
+            *solution,
```

Comment:
so this implicitly has to match the template in number of argument, and the position of the args is magical.